include/seqan3/alphabet/concept.hpp

```diff
@@ -14,24 +14,668 @@
 
 #include <iostream>
 
+#include <seqan3/alphabet/adaptation/char.hpp>
+#include <seqan3/alphabet/adaptation/uint.hpp>
+#include <seqan3/alphabet/concept_pre.hpp>
+#include <seqan3/alphabet/detail/member_exposure.hpp>
+#include <seqan3/alphabet/exception.hpp>
 #include <seqan3/core/concept/cereal.hpp>
 #include <seqan3/core/concept/core_language.hpp>
+#include <seqan3/core/detail/customisation_point.hpp>
+#include <seqan3/core/detail/reflection.hpp>
+#include <seqan3/core/metafunction/basic.hpp>
 #include <seqan3/core/metafunction/function.hpp>
-#include <seqan3/alphabet/adaptation/pre.hpp>
-#include <seqan3/alphabet/concept_pre.hpp>
-#include <seqan3/alphabet/detail/member_exposure.hpp>
 #include <seqan3/std/concepts>
 
+// ============================================================================
+// to_rank()
+// ============================================================================
+
+namespace seqan3::detail::adl::only
+{
+
+//!\brief Functor definition for seqan3::to_rank.
+struct to_rank_fn
+{
+private:
+    SEQAN3_CPO_IMPL(2, to_rank(v)                       )    // ADL
+    SEQAN3_CPO_IMPL(1, seqan3::adaptation::to_rank(v)   )    // customisation namespace
+    SEQAN3_CPO_IMPL(0, v.to_rank()                      )    // member
```

Comment:
Ok, I feel I am missing something here. So just to get it right. In order to avoid ambigious function call you add the priority_tag, as a parameter to the functions. Therefor you get unambigious overloads. The impl functions are unconstrained (neither per concept nor per SFINAE), but down there you call the impl function with `priority_tag<2>`. In this case you want to achieve, that it first tries the impl function with priority 2, if that fails then apply implicit conversion to priority_tag<1> etc. 
But if my type only supports .to_rank() and has no global to_rank in its namespace, would the compiler not emit an error in the first place?

Reply:
There is now an explicit requires clause.